### PR TITLE
Spread anywhere feature

### PR DESCRIPTION
**Pull request by butterunderflow**
Feature request from #670. 
The new feature will allow spread anywhere in a list creation expression.
For example, list creation like 
```rescript
let _ = list{1, 2, 3, ...x, 4, 5, ...y, 6}
```
will no longer trigger an error anymore, but parsed to
``` OCaml
let _ = Belt.List.concatMany [|(1 :: 2 :: 3 :: x);(4 :: 5 :: y);[6]|]
```

Can anyone help review this feature PR? Thanks ^v^

### CODE REVIEW

**Comment by cristianoc** [comment]
Would you add printing tests?

**Comment by butterunderflow** [comment]
What is `printing tests` stand for? Like printing some parse results related to this feature?

**Comment by cristianoc** [comment]
Add a file in tests/print or extend an existing one where list examples are.

It will tell you how things print, and whether printing is idempotent, so you don't get different results if you format twice.

**Comment by butterunderflow** [comment]
Thanks, I think I understand. I'll give it a try.

**Comment by cristianoc** [comment]
See how the syntax of normal lists is preserved by printing.
A similar mechanism is needed for this extension to lists.

**Comment by namenu** [comment]
tiny concern, but would it be better not allowing this expression at all ?

lists are not okay when manipulation is occurred at non-head position.

Once spreadings are used inadvertently, it would cause performance problem.

By now, list spreading guarantees constant time

**Comment by butterunderflow** [comment]
IMHO, it's still a useful feature. For some reason.
First, IMHO again, list concatenation is a little bit verbose at present. We have to write some prefixes of Module.
```rescript
Belt.List.concatMany([list{1, 2, ...x}, list{3, ...x}, list{4, ...x}, list{5}])
// or 
Belt.List.concat(list{1, 2, ...x}, list{3, ...x})
// or 
List.concat(list{list{1, 2, ...x}, list{3, ...x}})
```
Second, I think if we support @ operator (like OCaml), the verbose issues will be moderated. But for the existence of list open token `list{`,  looks still not concise enough. Like we have to tell the reader repeatedly "it's a **list**".
```rescript
// pretend we have operator @ 
list{1, 2, ...x} @ list{3, ...x} @ list{4, ...x} @ list{5}
list{1,2, ...x} @ list{3, ...x}
```
If we let spread show anywhere, we can just write:
```rescript
list{1, 2, ...x, 3, ...x, 4, ...x, 5}
list{1, 2, ...x, 3, ...x}
```
I think that's more concise, right?

And, as far as I know, literally, the word "spread" itself is not intuitively associated with constant time operation, unlike "prepend".

Finally, the performance issue you mentioned does exist while this feature is being misused. But this feature will not downside any performance of existing code, and I think (**I'm not sure**) concatenate seems can be optimized in some way. So I think it's still a good deal.  I'll always wait for more suggestions.

**Comment by IwanKaramazow** [comment]
Before adding this to the language, we should take a closer look at the design of this feature.
What is the reach of this problem? There are 3 vectors:
- volume of GH issues, forum posts, twitter etc…
- how blocking is this problem if the developer encounters
- how many of the total ReScript devs does it affect in their day to day work?

When making design decisions regarding new syntax in the language, it’s also important to weigh the costs, not all of which are performance, against the potential benefits. The following things can happen:
- The language gets larger.
- The language gets slower.
- The documentation becomes more extensive.
- We're spending time developing new syntax, rather than refining other parts of the language.
- We run the risk of introducing changes that don't play well with existing syntax. (Why don't we allow this for array then?)

**Comment by bobzhang** [comment]
Let me explain a bit why this feature it is useful and why you should not worry about performance.

This feature is very helpful for expressing lots of  algorithms, for example 

```re
Add(e1,e2) => list {... compile(e1), ... compile(e2), Add }
```
It is much more readable than the alternatives, and as efficient as you could, in theory the compiler can do 
a better job than you since it kows the `compile` is a temporay variable and can do some smart mutations.

So the argumets are:
- You should avoid list if you care about performance
- If you want to concate two lists, the alternative is not better at performance, you have to express this pattern, by
disabling this syntax does not mean you can avoid the problem, JavaScript is also consistent here

Before we add this feature, this used to be a bug, the syntax was accepted but not reported error, it is worse so 
that we are not enriching the syntax: we are **fixing** existing issues.

With regard to developer resources, this is the feature I would like to add on day one when making BuckleScript, 
finally, I have some dedicated resources to make it possible, it also makes the sample code for [PL course in ReScript](https://forum.rescript-lang.org/t/pl-course-in-rescript/3818) much more readable, instructions concatenation are everywhere. 

Hope my explanantions make the motivation clear

File: tests/parsing/errors/other/expected/spread.res.txt [hunk 1]
```diff
@@ -114,7 +100,7 @@ let arr = [|x;y|]
 let [|arr;_|] = [|1;2;3|]
 let record = { x with y }
 let { x; y } = myRecord
-let myList = x :: y
+let myList = Belt.List.concatMany [|x;y|]
```
Comment (bobzhang) [inline tests/parsing/errors/other/expected/spread.res.txt:103]: - this case should be specialized?
- this case too `{ ... x}`

File: tests/printer/expr/expected/list.res.txt [hunk 2]
```diff
@@ -2,6 +2,7 @@ let x = list{}
 let x = list{1}
 let x = list{1, 2}
 let x = list{1, 2, 3}
+let x = Belt.List.concatMany([list{1, 2, ...x}, list{3, ...x}])
```
Comment (bobzhang) [inline tests/printer/expr/expected/list.res.txt:6]: could this be improved? print it as `list{ 1, 2, ...x , 3 , ...x}`

Comment (butterunderflow) [reply tests/printer/expr/expected/list.res.txt:6]: It could be, but just by changing the printer can cause some problems. For example, list{. . x, . .y} and Belt.List.concatMany([x, y]) will be printed as the same thing. I am still trying to solve this problem.

Comment (bobzhang) [reply tests/printer/expr/expected/list.res.txt:6]: can you add an attribute to Belt.List.concatMany to avoid accidental simplifications?

Comment (cristianoc) [reply tests/printer/expr/expected/list.res.txt:6]: This is what is done for string templates, I believe, to differentiate from normal string concatenation.

Comment (cristianoc) [reply tests/printer/expr/expected/list.res.txt:6]: Check what `isTemplateLiteral` does.

Comment (butterunderflow) [reply tests/printer/expr/expected/list.res.txt:6]: Thank you for all your suggestions, I'll learn more about that.

**Comment by IwanKaramazow** [comment]
Thanks for the context @bobzhang 👍

File: src/res_parsetree_viewer.ml [hunk 3]
```diff
@@ -618,6 +618,26 @@ let isTemplateLiteral expr =
...
+let isSpreadBeltListConcat expr =
+  match expr.pexp_desc with
+  | Pexp_ident
+      {
+        txt =
+          Longident.Ldot
+            (Longident.Ldot (Longident.Lident "Belt", "List"), "concatMany");
+      }
+    when hasSpreadAttr expr.pexp_attributes ->
+    true
```
Comment (bobzhang) [inline src/res_parsetree_viewer.ml:638]: simplified: ` -> hasSpreadAttr exp.pexp_attributes`

File: tests/parsing/errors/other/expected/spread.res.txt [hunk 4]
```diff
@@ -42,79 +42,64 @@ Explanation: since records have a known, fixed shape, a spread like `{a, ...b}`
...
-   9 │ 
-  10 │ type t = {...a}
+  5 │ let {...x, ...y} = myRecord
+  6 │ 
+  7 │ let list{...x, ...y} = myList
+  8 │ 
+  9 │ type t = {...a}
 
   List pattern matches only supports one `...` spread, at the end.
 Explanation: a list spread at the tail is efficient, but a spread in the middle would create new lists; out of performance concern, our pattern matching currently guarantees to never create new intermediate data.
```
Comment (bobzhang) [inline tests/parsing/errors/other/expected/spread.res.txt:62]: we may need update the docs about pattern in later commits, it is not supported since it is not deterministic

File: tests/printer/expr/expected/list.res.txt [hunk 5]
```diff
@@ -2,6 +2,8 @@ let x = list{}
 let x = list{1}
 let x = list{1, 2}
 let x = list{1, 2, 3}
+let x = list{1, 2, ...x, 3, ...x}
+let x = Belt.List.concatMany([list{1, 2, ...x}, [list{3, ...x}]])
 
 let x = list{
```
Comment (bobzhang) [inline tests/printer/expr/expected/list.res.txt:8]: can you add some more test cases, some edge cases like:
```re
let x = list{ ...xs}
let x = list{1, ...xs}
let x = list{xs, ... ys}
let x = list{ ...xs, ... ys}
```

**Comment by bobzhang** [comment]
@butterunderflow looks good to me once the comments are addressed.
Note `Belt.List.concatMany` is an implementation detail, in theory, we could swap more efficient ones in the future

**Comment by butterunderflow** [comment]
Done. Please check again, thanks!

**Comment by bobzhang** [comment]
@butterunderflow Thanks, cannot wait for to use it in the course!

**Comment by cristianoc** [comment]
@butterunderflow would you add a line to CHANGELOG.md

**Comment by butterunderflow** [comment]
Done. Please check https://github.com/rescript-lang/syntax/pull/698
